### PR TITLE
98-port-densenet3d

### LINKED ISSUE #98: Port densenet
a subtask of #78, port densenet for 2d/3d classification

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #98 .

### Description
Implemented DenseNet3D network and unit test.
Already trained and verified locally with internal 3D classification dataset.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### CODE REVIEW

File: monai/networks/nets/densenet3d.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            if ft_module in k:
+                parameters.append({'params': v})
+                break
+        else:
+            parameters.append({'params': v, 'lr': 0.0})
+
+    return parameters
+
+
+class _DenseLayer(nn.Sequential):
```
Comment (ericspod) [inline monai/networks/nets/densenet3d.py]: Inheriting from Sequential and then providing your own forward is not compatible with torchscript I don't think. I would change this to inherit Module and have a Sequential member instead.

Comment (Nic-Ma) [reply monai/networks/nets/densenet3d.py]: Thanks @ericspod  for your review.
Welcome to change to your torchscript style later if you like.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

BTW, I also fixed an **out of memory** issue in UNet unit test in this PR.
If we don't disable auto_grad, PyTorch will compute grad and will not release memory after this test.
Then if run several networks, it will be out of memory.
Could you please help double-check it?
Thanks.

**Comment by ericspod** [comment]
How much memory was there on your test machine? Running all tests locally I see 500MB usage only. The UNet example doesn't run on GPU here anyway.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

The issue is not only about UNet, I mean if you **don't disable auto_grad** and run **10 networks continuously**(_including UNet_), it will fail with "out of memeory".
Thanks.

**Comment by ericspod** [comment]
Yes if you have 10 copies kept in memory that will happen but the unit tests shouldn't be doing that, the garbage collector does kick in. The no_grad is correct anyhow so the test change is fine.

**Comment by Nic-Ma** [comment]
> Yes if you have 10 copies kept in memory that will happen but the unit tests shouldn't be doing that, the garbage collector does kick in. The no_grad is correct anyhow so the test change is fine.

Cool, thanks for your review!

**Pull request merged**